contracts/HATVault.sol

```diff
@@ -0,0 +1,1109 @@
+    }
+
+    /**
+    * @notice Called by the vault's owner to set a pending request for the
+    * maximum percentage of the vault that can be paid out as a bounty.
+    * Cannot be called if there is an active claim that has been submitted.
+    * Max bounty should be less than or equal to `HUNDRED_PERCENT`.
+    * The pending value can later be set after the time delay (of 
+    * HATVaultsRegistry.GeneralParameters.setMaxBountyDelay) had passed.
+    * Max bounty should be less than or equal to `MAX_BOUNTY_LIMIT`
+    * @param _maxBounty The maximum bounty percentage that can be paid out
+    */
+    function setPendingMaxBounty(uint256 _maxBounty)
+    external
+    onlyOwner noActiveClaim {
+        if (_maxBounty > MAX_BOUNTY_LIMIT)
+            revert MaxBountyCannotBeMoreThanMaxBountyLimit();
+        pendingMaxBounty.maxBounty = _maxBounty;
+        // solhint-disable-next-line not-rely-on-time
+        pendingMaxBounty.timestamp = block.timestamp;
+        emit SetPendingMaxBounty(_maxBounty, pendingMaxBounty.timestamp);
+    }
+
+    /**
+    * @notice Called by the vault's owner to set the vault's max bounty to
+    * the already pending max bounty.
+    * Cannot be called if there are active claims that have been submitted.
+    * Can only be called if there is a max bounty pending approval, and the
+    * time delay since setting the pending max bounty had passed.
+    * Max bounty should be less than or equal to `MAX_BOUNTY_LIMIT`
+    */
+    function setMaxBounty() external onlyOwner noActiveClaim {
+        if (pendingMaxBounty.timestamp == 0) revert NoPendingMaxBounty();
+
+        HATVaultsRegistry.GeneralParameters memory generalParameters = registry.getGeneralParameters();
+
+        // solhint-disable-next-line not-rely-on-time
+        if (block.timestamp - pendingMaxBounty.timestamp <
+            generalParameters.setMaxBountyDelay)
+            revert DelayPeriodForSettingMaxBountyHadNotPassed();
+        maxBounty = pendingMaxBounty.maxBounty;
+        delete pendingMaxBounty;
+        emit SetMaxBounty(maxBounty);
+    }
+
+    /**
+    * @notice Called by governance to disable all deposits to the vault
+    * @param _depositPause Are deposits paused
+    */
+    function setDepositPause(bool _depositPause) external onlyOwner {
+        depositPause = _depositPause;
+        emit SetDepositPause(_depositPause);
+    }
+
+    /**
+    * @notice change the description of the vault
+    * only calleable by the owner of the vault
+    * @param _descriptionHash the hash of the vault's description.
+    */
+    function setVaultDescription(string memory _descriptionHash) external onlyOwner {
+        emit SetVaultDescription(_descriptionHash);
+    }
+
+    /**
+    * @notice Called by vault's owner to set the vault's reward controller
+    * @param _newRewardController The new reward controller
+    */
+    function setRewardController(IRewardController _newRewardController) external onlyOwner {
```

Comment:
One possible scenario with changing the rewardcontroller:
- the hats rewardcontroller HRC has a reward program for X
- X.rewardController = HRC
- halfway the reward program, X.rewardController changes to another 
- user balances will not be updated anymore
- rewards will be assigned on the basis of the current balance of shares or something (should check, but in any case it goes wrong :)

Reply:
Other potential problem is setting it to a value that breaks the transfer

Reply:
Hmmm right, but then what should we do about it? Remove changing the rewardController?

Reply:
For (a) the issue of detaching the rewardcontroller while payout is in process: we coudl do an explicit check (say that allocpont must be 0)

For (b) so the callback to the rewardController in vault._beforeTokenTransfer happens before any state changes if this is a simple transfer. Need to check if that is the case for all functions. In that case, that avoids reentrancy attacks. And we can do a try... catch around the updates so that the _beforeTokenTransfer does not revert.

Reply:
@ben-kaufman